src/ctx.rs

```diff
@@ -349,39 +345,29 @@ macro_rules! into_ctx_float_impl {
 into_ctx_float_impl!(f32, 4, super::Endian);
 into_ctx_float_impl!(f64, 8, super::Endian);
 
-#[inline(always)]
-fn get_str_delimiter_offset(bytes: &[u8], idx: usize, delimiter: u8) -> usize {
-    let len = bytes.len();
-    let mut i = idx;
-    let mut byte = bytes[i];
-    // TODO: this is still a hack and getting worse and worse - this hack has come from dryad -> goblin -> scroll :D
-    if byte == delimiter {
-        return i;
-    }
-    while byte != delimiter && i < len {
-        byte = bytes[i];
-        i += 1;
-    }
-    // we drop the terminator/delimiter unless we're at the end and the byte isn't the terminator
-    if i < len || bytes[i - 1] == delimiter {
-        i -= 1;
-    }
-    i
-}
-
 impl<'a> TryFromCtx<'a, StrCtx> for &'a str {
     type Error = error::Error;
     #[inline]
     /// Read a `&str` from `src` using `delimiter`
-    fn try_from_ctx(src: &'a [u8], StrCtx {delimiter}: StrCtx) -> Result<Self, Self::Error> {
-        let _len = src.len();
-        let delimiter_offset = get_str_delimiter_offset(src, 0, delimiter);
-        let count = delimiter_offset;
-        if count == 0 { return Ok("") }
-        let bytes = &src[..count];
-        str::from_utf8(bytes).map_err(| _err | {
-            error::Error::TooBig{size: count, len: src.len()}
-        })
+    fn try_from_ctx(src: &'a [u8], ctx: StrCtx) -> Result<Self, Self::Error> {
+        let len = match ctx {
+            StrCtx::Delimiter(delimiter) => src.iter().take_while(|c| **c != delimiter).count(),
+            StrCtx::DelimiterUntil(delimiter, len) => {
+                src
+                    .iter()
+                    .take_while(|c| **c != delimiter)
+                    .take(len)
+                    .count()
+            }
+            StrCtx::Length(len) => len,
+        };
+
+        if len > src.len() {
+            return Err(error::Error::TooBig{size: len, len: src.len()});
+        };
+
+        str::from_utf8(&src[..len])
+            .map_err(|_| error::Error::Custom("utf8 error".to_owned()))
```

Comment:
In original scroll I believe I had bad input with static string; we can't use custom here because it allocates and would force this entire impl to be behind a feature flag (also this won't compile now in `no_std`); I understand now what you mean by `From` being a bound, since it means we can't use std lib or core errors.

So we can either add an extra variant like original, or try out your generic version.